# Business Understanding

The company wants to increase revenue by recommending products
that customers are likely to purchase based on their previous purchases.

# Data Understanding

The dataset contains 100,000 transactions and 20 product groups.
Each ProdX column indicates whether that product group occurred
in the transaction.


In [13]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

df = pd.read_csv('drone_prod_groups.csv', sep=',')
print(df)

           ID  Prod1   Prod2   Prod3   Prod4   Prod5   Prod6   Prod7   Prod8  \
0           1      0       0       0       0       0       0       0       0   
1           2      0       1       0       0       0       0       0       0   
2           3      0       0       0       0       0       0       1       0   
3           4      1       0       0       1       0       0       0       0   
4           5      0       0       0       0       0       0       0       0   
...       ...    ...     ...     ...     ...     ...     ...     ...     ...   
99995   99996      0       0       1       0       0       0       0       0   
99996   99997      0       0       0       0       0       0       0       0   
99997   99998      0       1       0       0       0       0       0       0   
99998   99999      0       1       1       0       0       0       0       0   
99999  100000      0       0       0       0       1       0       0       0   

        Prod9  ...   Prod11   Prod12   

# Data Preparation

The transaction ID is removed and the product indicators are
converted to Boolean values.

In [14]:
# drop id
df = df.drop(columns='ID')

# replace 1.0 values with True
df = df.apply(lambda x: x.map(lambda y: True if y == 1.0 else False))
df



,Prod1,Prod2,Prod3,Prod4,Prod5,Prod6,Prod7,Prod8,Prod9,Prod10,Prod11,Prod12,Prod13,Prod14,Prod15,Prod16,Prod17,Prod18,Prod19,Prod20
0,False,False,False,False,False,False,False,False,True,False,False,False,False,False,True,False,False,False,False,True
1,False,True,False,False,False,False,False,False,True,False,False,False,False,False,True,True,True,True,True,True
2,False,False,False,False,False,False,True,False,False,False,False,False,False,False,False,False,False,False,True,True
3,True,False,False,True,False,False,False,False,False,False,True,False,False,False,False,False,False,False,True,True
4,False,False,False,False,False,False,False,False,True,False,False,False,False,False,True,False,False,False,True,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
99995,False,False,True,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False
99996,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,False,True,False,False
99997,False,True,False,False,False,False,False,False,True,False,False,False,False,False,True,False,False,True,False,False
99998,False,True,True,False,False,False,False,False,False,False,False,False,False,False,False,False,True,False,True,True


# Modeling

Apriori is used to find frequent itemsets

In [15]:
frequent_itemsets = apriori(df, min_support=0.1, use_colnames=True)
frequent_itemsets

,support,itemsets
0,0.10998,frozenset({Prod1})
1,0.13098,frozenset({ Prod2})
2,0.10459,frozenset({ Prod5})
3,0.13499,frozenset({ Prod7})
4,0.16179,frozenset({ Prod8})
5,0.19853,frozenset({ Prod9})
6,0.10848,frozenset({ Prod11})
7,0.15971,frozenset({ Prod12})
8,0.14557,frozenset({ Prod14})
9,0.11880,frozenset({ Prod15})


In [16]:
rules = association_rules(frequent_itemsets, metric='confidence', min_threshold=0.4)

rules = rules.sort_values(by='confidence', ascending=False)

rules

,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
0,frozenset({ Prod15}),frozenset({ Prod9}),0.11880,0.19853,0.11145,0.938131,4.725388,1.0,0.087865,12.954372,0.894663,0.541335,0.922806,0.749754
2,frozenset({ Prod20}),frozenset({ Prod19}),0.14798,0.20626,0.13476,0.910664,4.415125,1.0,0.104238,8.884845,0.907849,0.613997,0.887449,0.782007
3,frozenset({ Prod19}),frozenset({ Prod20}),0.20626,0.14798,0.13476,0.653350,4.415125,1.0,0.104238,2.457869,0.974508,0.613997,0.593144,0.782007
1,frozenset({ Prod9}),frozenset({ Prod15}),0.19853,0.11880,0.11145,0.561376,4.725388,1.0,0.087865,2.009011,0.983664,0.541335,0.502243,0.749754


# Evaluation

The rules are evaluated using support, confidence and lift.

The association rules show two strong product relationships: Prod15 → Prod9 and Prod20 → Prod19. Both have high confidence (93.8% and 91.1%) and high lift (4.73 and 4.42), meaning these products are frequently purchased together.

The support for Prod15 and Prod9 is 11.15%, meaning they appear together in 11.15% of all transactions. Prod20 and Prod19 have a higher support of 13.48%, so this combination occurs more frequently.